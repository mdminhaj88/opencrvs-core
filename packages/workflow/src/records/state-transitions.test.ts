/*
 * This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at https://mozilla.org/MPL/2.0/.
 *
 * OpenCRVS is also distributed under the terms of the Civil Registration
 * & Healthcare Disclaimer located at http://opencrvs.org/license.
 *
 * Copyright (C) The OpenCRVS Authors located at https://github.com/opencrvs/opencrvs-core/blob/master/AUTHORS.
 */
// mock getPractitionerLocations function from
// packages/workflow/src/features/user/utils.ts
// so it always returns

import {
  Location,
  Practitioner,
  REGISTERED_RECORD,
  SavedBundleEntry,
  Task
} from '@opencrvs/commons/types'
import { toCorrectionRequested } from './state-transitions'

const practitioner = REGISTERED_RECORD.entry.find(
  (entry): entry is SavedBundleEntry<Practitioner> =>
    entry.resource.resourceType === 'Practitioner'
)!.resource

const practitionerRole = REGISTERED_RECORD.entry.find(
  (entry): entry is SavedBundleEntry<Practitioner> =>
    entry.resource.resourceType === 'PractitionerRole'
)

jest.mock('@workflow/features/registration/fhir/fhir-utils', () => {
  const actual = jest.requireActual(
    '@workflow/features/registration/fhir/fhir-utils'
  )

  return {
    ...actual,
    getFromFhir: (suffix: string) => {
      if (suffix.startsWith('/Location/')) {
        const id = suffix.replace('/Location/', '')

        return REGISTERED_RECORD.entry.find(
          (entry): entry is SavedBundleEntry<Location> =>
            entry.resource.id === id
        )?.resource
      }
      switch (suffix) {
        case '/PractitionerRole?practitioner=7dc24408-d409-4ec0-bd9d-4a5aa817dcab':
          return {
            entry: [practitionerRole]
          }

        default:
          // eslint-disable-next-line no-console
          console.log('request not handled:', suffix)
          throw new Error('Not implemented')
      }
    }
  }
})

describe('functions that take record from one state to another', () => {
  describe('* -> correction requested', () => {
    it('should be defined', async () => {
      const correctionRequested = await toCorrectionRequested(
        REGISTERED_RECORD,
        practitioner,
        {
          requester: 'MOTHER',
          hasShowedVerifiedDocument: true,
          noSupportingDocumentationRequired: true,
          attachments: [],
          payment: {
            type: 'MANUAL',
            amount: 100,
            outcome: 'COMPLETED',
            date: '2021-08-10T10:00:00.000Z'
          },
          location: {
            _fhirID: 'Location/61e07903-bf3c-4bd9-9b20-6cf0e3bb982c'
          },
          values: [],
          reason: 'OTHER',
          otherReason: 'Other reason',
          note: 'Note'
        },
        []
      )

      const newResources = correctionRequested.entry
        .map((entry) => entry.resource)
        .filter(
          (resource) =>
            !REGISTERED_RECORD.entry.find(
              (entry) => entry.resource.id === resource.id
            )
        )

      const changedResources = correctionRequested.entry
        .map((entry) => entry.resource)
        .filter(
          (resource) =>
            resource.id &&
            !REGISTERED_RECORD.entry.find(
              (entry) => entry.resource === resource
            )
        ) as unknown as Task[]

      const withoutDynamicIdentifiers = changedResources.map((task) => {
        return {
          ...task,
          lastModified: 'STATIC',
          encounter: {
            reference: 'STATIC'
          },
          extension: task.extension.map((extension) => {
            if (
              extension.url !==
              'http://opencrvs.org/specs/extension/paymentDetails'
            ) {
              return extension
            }
            return {
              ...extension,
              valueReference: {
                reference: 'STATIC'
              }
            }
          })
        }
      })

      expect(newResources).toMatchSnapshot('new resources')
      expect(withoutDynamicIdentifiers).toMatchSnapshot('changed resources')
    })
  })
})
