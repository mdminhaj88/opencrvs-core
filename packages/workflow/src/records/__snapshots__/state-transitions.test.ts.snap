// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`functions that take record from one state to another * -> correction requested should be defined: changed resources 1`] = `
Array [
  Object {
    "businessStatus": Object {
      "coding": Array [
        Object {
          "code": "CORRECTION_REQUESTED",
          "system": "http://opencrvs.org/specs/reg-status",
        },
      ],
    },
    "code": Object {
      "coding": Array [
        Object {
          "code": "BIRTH",
          "system": "http://opencrvs.org/specs/types",
        },
      ],
    },
    "encounter": Object {
      "reference": "STATIC",
    },
    "extension": Array [
      Object {
        "url": "http://opencrvs.org/specs/extension/informants-signature",
        "valueString": "/ocrvs/c647d125-68ee-4794-b320-d472aadd8c09.png",
      },
      Object {
        "url": "http://opencrvs.org/specs/extension/contact-person-phone-number",
        "valueString": "+260745645645",
      },
      Object {
        "url": "http://opencrvs.org/specs/extension/contact-person-email",
        "valueString": "informant@example.com",
      },
      Object {
        "url": "http://opencrvs.org/specs/extension/timeLoggedMS",
        "valueInteger": 0,
      },
      Object {
        "url": "http://opencrvs.org/specs/extension/paymentDetails",
        "valueReference": Object {
          "reference": "STATIC",
        },
      },
      Object {
        "url": "http://opencrvs.org/specs/extension/noSupportingDocumentationRequired",
        "valueBoolean": true,
      },
      Object {
        "url": "http://opencrvs.org/specs/extension/requestingIndividual",
        "valueString": "MOTHER",
      },
      Object {
        "url": "http://opencrvs.org/specs/extension/hasShowedVerifiedDocument",
        "valueBoolean": true,
      },
      Object {
        "url": "http://opencrvs.org/specs/extension/regLastUser",
        "valueReference": Object {
          "reference": "Practitioner/7dc24408-d409-4ec0-bd9d-4a5aa817dcab",
        },
      },
      Object {
        "url": "http://opencrvs.org/specs/extension/regLastLocation",
        "valueReference": Object {
          "reference": "Location/d8d16138-1d3a-42e7-bf85-d7f3f657c5d3",
        },
      },
      Object {
        "url": "http://opencrvs.org/specs/extension/regLastOffice",
        "valueReference": Object {
          "reference": "Location/61e07903-bf3c-4bd9-9b20-6cf0e3bb982c",
        },
        "valueString": "Ibombo District Office",
      },
    ],
    "focus": Object {
      "reference": "Composition/c5811d36-934d-40f9-94b4-15194d562e45",
    },
    "id": "8dae3f40-70f8-4463-aae1-fd56aec0c8ee",
    "identifier": Array [
      Object {
        "system": "http://opencrvs.org/specs/id/draft-id",
        "value": "03d087a0-1e20-4443-aca3-d48461d4d597",
      },
      Object {
        "system": "http://opencrvs.org/specs/id/birth-tracking-id",
        "value": "BF6W25K",
      },
      Object {
        "system": "http://opencrvs.org/specs/id/birth-registration-number",
        "value": "2023BF6W25K",
      },
    ],
    "input": Array [],
    "intent": "proposal",
    "lastModified": "STATIC",
    "note": Array [
      Object {
        "text": "Note",
      },
    ],
    "reason": Object {
      "extension": Array [
        Object {
          "url": "http://opencrvs.org/specs/extension/otherReason",
          "valueString": "Other reason",
        },
      ],
      "text": "OTHER",
    },
    "requester": Object {
      "agent": Object {
        "reference": "Practitioner/7dc24408-d409-4ec0-bd9d-4a5aa817dcab",
      },
    },
    "resourceType": "Task",
    "status": "requested",
  },
]
`;

exports[`functions that take record from one state to another * -> correction requested should be defined: new resources 1`] = `
Array [
  Object {
    "detail": Array [
      Object {
        "amount": Object {
          "value": 100,
        },
        "date": "2021-08-10T10:00:00.000Z",
        "type": Object {
          "coding": Array [
            Object {
              "code": "MANUAL",
            },
          ],
        },
      },
    ],
    "outcome": Object {
      "coding": Array [
        Object {
          "code": "COMPLETED",
        },
      ],
    },
    "resourceType": "PaymentReconciliation",
    "status": "active",
    "total": Object {
      "value": 100,
    },
  },
  Object {
    "resourceType": "Encounter",
    "status": "finished",
  },
]
`;
